Updates to the Bayesian Regression tutorial

neerajprad opened the pull request:
Addresses #2083, #2054. 

This updates the Bayesian Regression tutorial to:
 - Remove usage of `random_module` and `TracePredictive` in favor of `PyroModule` and `Predictive`.
 - Adds a section on model serving using `TorchScript`.
 - Mostly other small changes for clarification.

nbviewer [link](https://nbviewer.jupyter.org/github/pyro-ppl/pyro/blob/43a5f9e3b7057f066c42bcb301efdb1be34b576e/tutorial/source/bayesian_regression.ipynb)

fehiepsi commented on tutorial/source/bayesian_regression.ipynb:
```diff
@@ -92,333 +109,367 @@
+    "                ax=ax[1])\n",
     "ax[1].set(xlabel=\"Terrain Ruggedness Index\",\n",
     "          ylabel=\"log GDP (2000)\",\n",
-    "          title=\"Afri
```
We don't need to register for unconstrained parameters. The original `nn.Parameter` also works.

fehiepsi replied:
This comment is for the definition of `LinearRegression` class. I don't know why it displays a different position here.

neerajprad replied:
You mean we can use `nn.Parameter` instead of `PyroParam`? That's true, but I wanted to write something that would work with constraints too. I'm happy to remove it if it seems confusing.

neerajprad replied:
Done.

fehiepsi commented on tutorial/source/bayesian_regression.ipynb:
```diff
@@ -92,333 +109,367 @@
+    "                ax=ax[1])\n",
     "ax[1].set(xlabel=\"Terrain Ruggedness Index\",\n",
     "          ylabel=\"log GDP (2000)\",\n",
-    "          title=\"Afri
```
not important: I think you can also define `sigma` as PyroSample.

fehiepsi replied:
This is for the definition of `BayesianRegression` class.

neerajprad replied:
That's right, I didn't do that because we don't need to set it as an attribute, and I just wanted to show that we can call regular pyro primitives in the code too, but I should call this out probably.

neerajprad replied:
Actually `PyroSample` won't work here.

fehiepsi replied:
I meant to define `self.sigma = PyroSample(dist.Uniform(0., 10.))` at the module construction. I think that it is a part of the Bayesian NN, so considering it an attribute of BayesianRegression module makes sense to me. Anyway, this is just a minor point to me. :)

neerajprad replied:
That seems reasonable to me, will change.

neerajprad replied:
Actually, one issue that I see with this pattern is that it is easy to forget that `self.sigma` is a special attribute and we need to access it in the correct context. The current model is:

```python
    def forward(self, x, y=None):
        sigma = pyro.sample("sigma", dist.Uniform(0, 10))
        mean = self.linear(x).squeeze(-1)
        with pyro.plate("data", x.shape[0]):
            obs = pyro.sample("obs", dist.Normal(mean, sigma), obs=y)
        return mean
```

if we change it to have `self.sigma = PyroSample(dist.Uniform(1, 10))` in `__init__` and change the forward method to the following, 

```python
    def forward(self, x, y=None):
        mean = self.linear(x).squeeze(-1)
        with pyro.plate("data", x.shape[0]):
            obs = pyro.sample("obs", dist.Normal(mean, self.sigma), obs=y)
        return mean
```
we'll silently do the unintended thing of actually sampling `sigma` that has a batch size equal to the data, since it is easy to forget that `self.sigma` needs to be accessed in the correct context, i.e. we'll need to add `sigma = self.sigma` outside the plate so as to trigger `getattr` outside the plate context. I'm happy to do this latter thing, but that will need a little bit more explaining in the notebook as to why we did that. I'll make the change based on feedback.

fehiepsi replied:
Nice observation! This is easy to make mistakes. Thanks for pointing it out!

fehiepsi commented on tutorial/source/bayesian_regression.ipynb:
```diff
@@ -92,333 +109,367 @@
+    "                ax=ax[1])\n",
     "ax[1].set(xlabel=\"Terrain Ruggedness Index\",\n",
     "          ylabel=\"log GDP (2000)\",\n",
-    "          title=\"Afri
```
Is it your intention to reuse the learned `linear_reg_model`?

fehiepsi replied:
This comment is for the line `model = BayesianRegression(linear_reg_model)` in Using AutoGuide section.

neerajprad replied:
That was just to initialize the parameters to the ML estimates; maybe I should write a line to clarify that.

fehiepsi commented on tutorial/source/bayesian_regression.ipynb:
```diff
@@ -92,333 +109,367 @@
+    "                ax=ax[1])\n",
     "ax[1].set(xlabel=\"Terrain Ruggedness Index\",\n",
     "          ylabel=\"log GDP (2000)\",\n",
-    "          title=\"Afri
```
nit: `print(name, pyro.param(name).data)` will make the output cleaner

fehiepsi replied:
This is for the cell numbered 12.

fehiepsi commented on tutorial/source/bayesian_regression.ipynb:
```diff
@@ -623,64 +678,194 @@
+   "source": [
+    "## Model Serving via TorchScript\n",
+    "\n",
+    "Finally, note that the `model`, `guide` and the `predictive` utility class are all `torch.nn.Module` instances, and can be serialized as [TorchScript](https://pytorch.org/docs/stable/jit.html). \n",
```
typo: `predictive` -> `Predictive`

martinjankowiak commented:
generally lgtm. comments:
- "Each time linear_module is called,"  => add "... the forward method..." ?
- you're using the data twice if you set the priors using the MLE estimate?
- `num_samples=1000` but not vectorized? isn't that super slow? incidentally, i thought we had `num_particles` everywhere....
- this isn't relevant to this tutorial but why aren't the returned tensors in `quantiles` detached?
- strange language: "Refer to the Model Serving section for an inside look at how you can roll one yourself"
- "Note that for lower values of ruggedness there are many more data points, and as such, lesser wiggle room for the line of best fit" this isn't right. the coefficients are global. the band gets thicker because weights are multiplied by inputs

neerajprad commented:
> "Each time linear_module is called," => add "... the forward method..." ?

Done.

> you're using the data twice if you set the priors using the MLE estimate?

Yeah, this is problematic. I wanted to mention that we could do a better initialization for the autoguide mean if needed, but this is using the dataset twice. It will be best to just initialize using some sensible priors instead.

> num_samples=1000 but not vectorized? isn't that super slow? incidentally, i thought we had num_particles everywhere....

I should remove the `num_samples` argument. Its from the earlier code, and is only used with `TracePredictive`. In any case, models with NNs and `.squeeze`s don't always work with our vectorization. 

> this isn't relevant to this tutorial but why aren't the returned tensors in quantiles detached?

Since `nn.Parameter` has `requires_grad=True`, we end up with tensors with `gradient_fn` whenever the guide is called. I think this would have been the case earlier too, but I'm looking into this separately.

> strange language: "Refer to the Model Serving section for an inside look at how you can roll one yourself"

Fixed.

> "Note that for lower values of ruggedness there are many more data points, and as such, lesser wiggle room for the line of best fit" this isn't right. the coefficients are global. the band gets thicker because weights are multiplied by inputs

Yes, that's plain wrong. Fixed.

neerajprad commented on tutorial/source/bayesian_regression.ipynb:
```diff
@@ -16,7 +16,21 @@
+    " - [Linear Regression](#Linear-Regression)\n",
+    "   - [Training with PyTorch Optimizers](#Training-with-PyTorch-Optimizers)\n",
+    "   - [Regression Fit](#Plotting-the-Regression-Fit)\n",
+    " - [Bayesian Regression with Pyro's SVI](#Bayesian-Regression-with-Pyro's-Stochastic-Variational-Inference-(SVI))\n",
```
Fixed.

neerajprad commented:
I completely forgot that now that we have `guide` as a module, we can simply turn off the `requires_grad` attribute to False on all parameters by doing `guide.requires_grad_(False)`. With that we don't end up computing `grad_fn` and the output is much cleaner.

fritzo commented on tutorial/source/bayesian_regression.ipynb:
```diff
@@ -92,337 +109,372 @@
+    "                ax=ax[1])\n",
     "ax[1].set(xlabel=\"Terrain Ruggedness Index\",\n",
     "          ylabel=\"log GDP (2000)\",\n",
-    "          title=\"Afri
```
Are you sure this is necessary? I would have thought that we could simply
```py
class LinearRegression(nn.Linear, PyroModule):
    pass
```

neerajprad replied:
I was using `PyroParam` earlier, but since we are using `nn.Parameter` now, I think we can remove this.

fritzo commented:
Looks great!
